pallets/collator-assignment/src/lib.rs

```diff
@@ -159,16 +169,110 @@ pub mod pallet {
     }
 
     impl<T: Config> Pallet<T> {
+        pub(crate) fn order_paras_with_core_config(
+            mut bulk_paras: Vec<ChainNumCollators>,
+            mut pool_paras: Vec<ChainNumCollators>,
+            core_allocation_configuration: &CoreAllocationConfiguration,
+            target_session_index: T::SessionIndex,
+            number_of_collators: u32,
+            collators_per_container: u32,
+            collators_per_parathread: u32,
+        ) -> (Vec<ChainNumCollators>, bool) {
+            let core_count = core_allocation_configuration.core_count;
+            let max_number_of_bulk_paras = core_allocation_configuration
+                .max_parachain_percentage
+                .mul(core_count);
+
+            let enough_cores_for_bulk_paras = bulk_paras.len() <= max_number_of_bulk_paras as usize;
+
+            // Currently, we are sorting both bulk and pool paras by tip, even when for example
+            // only number of bulk paras are restricted due to core availability since we deduct tip from
+            // all paras.
+            // We need to sort both separately as we have fixed space for parachains at the moment
+            // which means even when we have some parathread cores empty we cannot schedule parachain there.
+            // We need to change this once other part of algorithm start to differentiate between
+            // bulk and pool paras while deducting the tip amount.
+            if !enough_cores_for_bulk_paras {
+                bulk_paras.sort_by(|a, b| {
+                    T::CollatorAssignmentTip::get_para_tip(b.para_id)
+                        .cmp(&T::CollatorAssignmentTip::get_para_tip(a.para_id))
+                });
+
+                pool_paras.sort_by(|a, b| {
```

Comment:
We are sorting `pool_paras` because since we deduct tip from all paras down the line. So, if we let's say do not sort `pool_paras` here, in collator assignment algorithm even the pool paras who paid for higher tip will still have lower priority than pool para with lower tip.

Reply:
We can change this later on when we differentiate between bulk and pool paras with regards to tip. I think @girazoki already opened an issue signifying that.

Reply:
But later in `order_paras` we sort all chains if `enough_collators_for_all_chain == false`. So unless I'm missing something, we can remove this and parathreads will still be sorted when needed? Or is there a case where `enough_collators_for_all_chain == true` and we need to charge the priority fee?

Reply:
The problem is we charge tip from all the paras, so if I do not sort pool paras here and `enough_collators_for_all_chain == true`, (which can happen when we have enough cores but not enough collators.), we will still be deducting tip from all paras but pool paras who paid higher tip still cannot be ahead of the pool para paying lower tip which I think is unfair.

Reply:
This will be solved by re-writing tip charging mechanism to differentiate between bulk and pool paras.

Reply:
Yeah I agree. roadmap

Reply:
yes indeed

Reply:
roadmap